walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,159 @@
+    if (events.isEmpty()) {
+      if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        // return empty arraylist
+         return eventsAsTimeRange;
+      } else {
+         eventsAsTimeRange.add(TimeRange.fromStartDuration(0, 1440));
+         return eventsAsTimeRange;
+      }
+    }
+
+    //filter out events that don't involve required attendees
+    List<Event> filteredEvents = filterEvents(events, request);
+
+    // make list of event objects in order to get attendees later
+    List<Event> eventsAsSortedList = getSortedTimeRanges(filteredEvents);
+
+    //flatten events to TimeRanges representing when any required person is unavailable
+    List<TimeRange> unavailableTimeRanges = getUnavailableTimeRanges(eventsAsSortedList);
+
+    //use unavailableTimeRanges to find available time ranges
+    // List<TimeRange> availableTimeRanges = getAvailableTimeRanges(unavailableTimeRanges);
+
+    throw new UnsupportedOperationException("TODO: continue building schedule");
+  }
+
+  public List<Event> filterEvents(Collection<Event> events, MeetingRequest request) {
+    List<Event> filteredEvents = new ArrayList<Event>();
+    Collection<String> requiredAttendees = request.getAttendees();
+
+    //not the most efficient way, but it works
+    for (Event event: events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      for (String attendee: eventAttendees) {
+        if (requiredAttendees.contains(attendee)) {
+          filteredEvents.add(event);
+        }
+      }
+      
+    }
+    return filteredEvents;
+  }
+
+  public List<Event> getSortedTimeRanges(List<Event> events) {
+    // Collections.sort() only takes in List, so must convert events to a list
+    List<Event> eventsAsList= new ArrayList<Event>();
+    for (Event event: events) {
+      eventsAsList.add(event);
+    }
+    Collections.sort(eventsAsList, Event.ORDER_BY_START);
+    return eventsAsList;
+  }
+
+  public List<TimeRange> getUnavailableTimeRanges(List<Event> events) {
+    List<TimeRange> unavailableTimeRanges = new ArrayList<TimeRange>();
+    List<TimeRange> eventsAsTimeRanges = new ArrayList<TimeRange>();
+
+    for (Event event: events) {
+      eventsAsTimeRanges.add(event.getWhen());
+    }
+
+    //case 1: events that do not overlap
+    for (TimeRange timeRange_1: eventsAsTimeRanges) {
+      boolean timeRangesOverlap = false;
+      for (TimeRange timeRange_2: eventsAsTimeRanges)  {
+        if (!(timeRange_1.equals(timeRange_2))) {
+          if ((timeRange_1.overlaps(timeRange_2))) {
+            timeRangesOverlap = true;
+          }
+        }
+      }
+      if (timeRangesOverlap == false) {
+         // ensure no double-counting timeRanges
+        if (!(unavailableTimeRanges.contains(timeRange_1))){
+          unavailableTimeRanges.add(timeRange_1);
+        }
+      }
+    }
+
+    //very brute-force way to cover overlapping time ranges
+    for (TimeRange timeRange_1: eventsAsTimeRanges) {
```

Comment:
i think in java, we typically don't use underscores for variables that are not constants. probably better to rename to timeRange1 or a more descriptive name. ditto for timeRange_2.